pymc/backends/zarr.py

```diff
@@ -452,7 +453,12 @@ def init_trace(
         )
         self.vars = [var for var in vars if var.name in self.varnames]
 
-        self.fn = model.compile_fn(self.vars, inputs=model.value_vars, on_unused_input="ignore")
+        self.fn = model.compile_fn(
+            self.vars,
+            inputs=model.value_vars,
+            on_unused_input="ignore",
+            borrow_vars=True,
+        )
```

Comment:
Did I mess it up. Doesn't `BaseTrace.__init__` already create the fn you need?

Reply:
All my comments about borrow and trust_input were based on this assumption. Note the base class does not call `model.compile_fn`, it sidesteps it completely, hence why I was surprised you were changing it previously

Reply:
No, the `BaseTrace` for each chain (`ZarrChain`) is initialized in `ZarrTrace`. Now I'm sending off the `fn` that is compiled in `ZarrTrace` over to every chain instance to use it without having to recompile it.

Reply:
We should do that for MultiTrace as well. I had that in mind as a follow up in #7578 where a lot of the wall time was just avoiding needless function compilations. 

My early comment still stands you shouldn't call `model.compile_fn` and you should (unless you don't want the speedup for some reason) have trust_input and borrow inputs and outputs to avoid deep copies needlessly. AFAICT the same logic should be applicable to the new and old traces.

Reply:
The reason I didn't do it for multiprocessing though is that we have to be careful about functions with RNGs, we actually need to copy the function and provide new shared RNGs for each function. If you're sharing `self.fn` across multiple chains (and they are called in different processes) the same concern applies here?

Reply:
Here is a related issue: https://github.com/pymc-devs/pymc/issues/7588

Reply:
If you don't copy the function and do the RNG swap it's better not to share for now, as it may mess up the RNG updates and you get crappy random draws.

Reply:
Here is the kind of logic needed for this function copy and RNG swap: https://github.com/pymc-devs/pymc-extras/blob/c1809e8149fc89ac6eadf4bf73050ea6fe82955c/pymc_experimental/sampling/optimizations/conjugate_sampler.py#L84-L98

Reply:
Is the problem that the random generators inside `self.fn` are shared across chains? What I mean is that, is the problem that when different chains call the same `fn`, the random generator state will advance mixing up the different chains? I can see that this could happen in the context of a single process that runs multiple chains, either in threads or sequentially. But I don't see this happening in different processes.
Or is the problem that we don't manage to control the random generator state that gets called in `fn` as we do with step methods? If this is the real problem, we could address it using the `RandomGeneratorState`, and `random_generator_from_state` and `get_state_from_generator` to create copies of random generators that adhere to whatever state was inputed.

Reply:
Anyway, I think that #7588 really captures the core problem with what might be going on, so I don't think that it's worth opening a different issue. This PR doesn't close the issue though, but I don't think that it's intention should be to do so. I have an idea to try and tackle 7588 using the `RandomGeneratorState` stuff that I developed here.